antonbabenko/pre-commit-terraform
# feat(`terraform_docs`): Add `terraform-docs` default markers support and describe how to migrate to them

Put an `x` into the box if that apply:


### Description of your changes

Allow users to switch to the standard insertion markers used by terraform-docs, to enable CI jobs to check that docs have been updated.

### How can we test changes
Based on this suggestion for testing hooks: https://github.com/antonbabenko/pre-commit-terraform/pull/25#issuecomment-431649551

Add the following to the pre-commit config:

```yaml
- repo: local
  hooks:
  - id: terraform_docs
    name: terraform_docs
    entry: ./hooks/terraform_docs.sh
    args:
    - --hook-config=--path-to-file=README.md
    - --hook-config=--use-standard-markers=false
    - --hook-config=--create-file-if-not-exist=true
    - --hook-config=--add-to-existing-file=true
    language: system
    files: (\.tf)$
```

Add a new file `outputs.tf`:
```hcl
output "my_output" {
  value       = aws_api_gateway_stage.blah.invoke_url
  description = "Description of the test output"
}
```
Stage the pre-commit config and the new outputs.tf file: `git add .pre-commit-config.yaml outputs.tf`

Using the existing README.md as a target:
- Test default behaviour has not changed:
  - `pre-commit run -a` should add the terraform docs between the existing markers
  - Delete the README. Then `pre-commit run -a` should create the file as normal
  - Delete the contents of the README. Then `pre-commit run -a` should add the terraform docs as normal
- Test the behaviour with "standard" markers
  - update the pre-commit config to set `--hook-config=--use_standard_markers=true`, and `git add -u`
  - Delete the README. Then `pre-commit run -a` should create the file as normal, but with different markers
  - `git add README.md`
  - update using terraform-docs: `terraform-docs markdown --output-file=README.md .`
  - there should be no changes made by terraform docs (`git diff`)
  - repeat this process in reverse (terraform-docs, then pre-commit)

For completeness, repeat the standard markers test for insertion between existing markers, and appending to an existing file with no markers (or check by code inspection).
